ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/FunctionArgumentsUtil.java

```diff
@@ -0,0 +1,227 @@
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.SqlArgument;
+import io.confluent.ksql.schema.ksql.types.SqlLambda;
+import io.confluent.ksql.schema.ksql.types.SqlLambdaResolved;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.Pair;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public final class FunctionArgumentsUtil {
+
+  private FunctionArgumentsUtil() {
+
+  }
+
+  public static class FunctionArgumentsAndContext {
+    private final List<TypeContext> contexts;
+    private final List<SqlArgument> arguments;
+    private final SqlType returnType;
+    private final KsqlScalarFunction function;
+
+    public FunctionArgumentsAndContext(
+        final List<TypeContext> contexts,
+        final List<SqlArgument> arguments,
+        final SqlType returnType,
+        final KsqlScalarFunction function
+    ) {
+      this.contexts = contexts;
+      this.arguments = arguments;
+      this.returnType = returnType;
+      this.function = function;
+    }
+
+    public List<TypeContext> getContexts() {
+      return contexts;
+    }
+
+    public List<SqlArgument> getArguments() {
+      return arguments;
+    }
+
+    public SqlType getReturnType() {
+      return returnType;
+    }
+
+    public KsqlScalarFunction getFunction() {
+      return function;
+    }
+  }
+
+  /**
+   * Given a function call node, we have to do a two pass processing of the 
+   * function arguments in order to properly handle any potential lambda functions.
+   * In the first pass, if there are lambda functions, we create a SqlLambda that only contains
+   * the number of input arguments for the lambda. We pass this first argument list
+   * to UdfFactory in order to get the correct function. We can make this assumption 
+   * due to Java's handling of type erasure (Function(T,R) is considered the same as
+   * Function(U,R)).
+   * In the second pass, we use the LambdaType inputTypes field to construct SqlLambdaResolved
+   * that has the proper input type list and return type. We also need to construct a list of
+   * contexts that should be used when processing each function argument subtree.
+   *
+   * @param expressionTypeManager an expression type manager
+   * @param functionCall  the function expression
+   * @param udfFactory  udf factory for the function in the expression
+   * @param context a type context
+   *
+   * @return a wrapper that contains a list of function arguments 
+   *         (any lambdas are SqlLambdaResolved), the ksql function, 
+   *         type contexts for use in further processing the function 
+   *         argument child nodes, and the return type of the ksql function
+   */
+  // CHECKSTYLE_RULES.OFF: CyclomaticComplexity
+  public static FunctionArgumentsAndContext getLambdaContextAndType(
```

Comment:
nit: naming - `getFunctionTypeInfo`? - it's really returning you all the info about the function argument types and return types, not just for lambdas